# PROJECT 3: MADELON DATA ANALYSIS

Author: William Buck

---

In [1]:
#!conda install psycopg2 --yes

In [36]:
%run madelon_analyzer.py
%run pickle_jar.py
%run Expanded_madelon_dataframes_and_features.py
%matplotlib inline

# Big Data - EDA

Below is EDA of the expanded Madelon dataset with 1000 features and 200,000 observations. To start with, I've imported 3000 rows and will be performing the same unsupervised learning techniques I used for the smaller UCI dataset to find and remove the noise.


In [3]:
# in this cell, I'm going to try and get a representative sample of the population
# https://www.surveymonkey.com/mp/sample-size-calculator/ (confidence level = 90%, margin of error = 1)

# con_6506, cur_6506 = con_cur_to_class_db()
# cur_6506.execute('SELECT * FROM madelon ORDER BY random() LIMIT 6506')
# results_6506 = cur_6506.fetchall()
# con_6506.close()

In [4]:
# # Dataframe of just 6506 samples for initial testing.

# samp_6506_bd_df = pd.DataFrame(results_6506)
# pd.to_pickle(samp_6506_bd_df, '6506_sample_big_data.p')

In [5]:
# samp_6506_bd_df.index = samp_6506_bd_df['_id']
# samp_6506_bd_df = samp_6506_bd_df.iloc[:,1:]

In [6]:
# # Dataframe of just 3000 samples for initial testing.

# con, cur = con_cur_to_class_db()
# %time cur.execute('SELECT * FROM madelon ORDER BY random() LIMIT 3000')
# %time results = cur.fetchall()
# con.close()

# # Dataframe of just 3000 samples for initial testing.

# samp_3000_bd_df = pd.DataFrame(results)

# pd.to_pickle(samp_3000_bd_df, '3000_sample_big_data.p')

In [7]:
X = samp_3000_bd_df.iloc[:200,1:-1]
X_2 = samp_3000_bd_df.iloc[200:400,1:-1]
X_3 = samp_3000_bd_df.iloc[400:600,1:-1]
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Columns: 1000 entries, feat_000 to feat_999
dtypes: float64(1000)
memory usage: 1.5 MB


In [8]:
# From this, I can tell that the big uci dataset is already normal
X.describe().T.mean()

count    200.000000
mean       0.001668
std        1.019437
min       -2.800271
25%       -0.681074
50%        0.001271
75%        0.682078
max        2.826687
dtype: float64

In [9]:
y = samp_3000_bd_df.iloc[:200,-1:]
y_2 = samp_3000_bd_df.iloc[200:400,-1:]
y_3 = samp_3000_bd_df.iloc[400:600,-1:]

In [10]:
# Making three sample test dataframes

X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=42, test_size=0.333)
X_2_train, X_2_test, y_2_train, y_2_test = train_test_split(X_2,y_2, random_state=42, test_size=0.333)
X_3_train, X_3_test, y_3_train, y_3_test = train_test_split(X_3,y_3, random_state=42, test_size=0.333)

train_df = pd.merge(X_train, y_train, left_index=True, right_index=True)
train_2_df = pd.merge(X_2_train, y_2_train, left_index=True, right_index=True)
train_3_df = pd.merge(X_3_train, y_3_train, left_index=True, right_index=True)

test_df = pd.merge(X_test, y_test, left_index=True, right_index=True)
test_2_df = pd.merge(X_2_test, y_2_test, left_index=True, right_index=True)
test_3_df = pd.merge(X_3_test, y_3_test, left_index=True, right_index=True)

In [14]:
# # NOTE: cell takes 7 minutes to run

# ma = madelon_analyzer(train_df, test_df)
# ma_2 = madelon_analyzer(train_2_df, test_2_df)
# ma_3 = madelon_analyzer(train_3_df, test_3_df)

# %time r2_bd_df = ma.mean_r2_for_all_features(train_df, DecisionTreeRegressor)

# r2_bd_df = r2_bd_df.sort_values('r2_score', ascending=False)

In [12]:
# pd.to_pickle(r2_bd_df, 'mean_r2_bd_sample_1.p')
r2_bd_df = pd.read_pickle('Data/mean_r2_bd_sample_1.p')
r2_bd_df.head(20)

r2_score   feature
956  0.876759  feat_956
639  0.843671  feat_639
315  0.601526  feat_315
269  0.513198  feat_269
341  0.509998  feat_341
724  0.483858  feat_724
867  0.477938  feat_867
701  0.473653  feat_701
336  0.465809  feat_336
736  0.458448  feat_736
395  0.448664  feat_395
769   0.43618  feat_769
526   0.30044  feat_526
920   0.27338  feat_920
504  0.229805  feat_504
681  0.220966  feat_681
257  0.167354  feat_257
829  0.141994  feat_829
308  0.121009  feat_308
808 -0.242516  feat_808

In [15]:
# %time r2_bd_2_df = ma_2.mean_r2_for_all_features(train_2_df, DecisionTreeRegressor)
# r2_bd_2_df = r2_bd_2_df.sort_values('r2_score', ascending=False)
# pd.to_pickle(r2_bd_2_df, 'mean_r2_bd_sample_2.p')
r2_bd_2_df = pd.read_pickle('Data/mean_r2_bd_sample_2.p')
r2_bd_2_df[:25]

r2_score   feature
639   0.845707  feat_639
956   0.808271  feat_956
315   0.575084  feat_315
867   0.485985  feat_867
341   0.462097  feat_341
701   0.417449  feat_701
724   0.409274  feat_724
526   0.347879  feat_526
769   0.327253  feat_769
829   0.314855  feat_829
269   0.311286  feat_269
336   0.293929  feat_336
920   0.290956  feat_920
395   0.284413  feat_395
257   0.279773  feat_257
808   0.211217  feat_808
736   0.200698  feat_736
504   0.198587  feat_504
681   0.188266  feat_681
308  0.0590268  feat_308
859    -0.5914  feat_859
33   -0.619932   feat_33
875  -0.662935  feat_875
138  -0.666162  feat_138
448  -0.667927  feat_448

In [16]:
# %time r2_bd_3_df = ma_3.mean_r2_for_all_features(train_3_df, DecisionTreeRegressor)
# r2_bd_3_df = r2_bd_3_df.sort_values('r2_score', ascending=False)
# pd.to_pickle(r2_bd_3_df, 'mean_r2_bd_sample_3.p')
r2_bd_3_df = pd.read_pickle('Data/mean_r2_bd_sample_3.p')
r2_bd_3_df[:25]

r2_score   feature
639   0.806627  feat_639
956   0.754375  feat_956
315   0.573237  feat_315
724   0.519356  feat_724
867   0.475584  feat_867
336   0.472721  feat_336
269   0.467021  feat_269
701   0.390734  feat_701
920   0.375489  feat_920
681   0.315068  feat_681
395   0.291377  feat_395
736   0.255419  feat_736
257   0.250702  feat_257
829    0.23143  feat_829
341   0.182425  feat_341
769   0.176203  feat_769
504   0.172482  feat_504
808   0.124137  feat_808
308 -0.0130461  feat_308
526  -0.329475  feat_526
824  -0.564693  feat_824
1    -0.661999    feat_1
693  -0.684159  feat_693
283  -0.715089  feat_283
788  -0.728706  feat_788

In [41]:
# First 20 features are all the same for the first 3 samples of the big madelon dataset

first_3_bd_samps_correlated_feats_list = \
list(set(list(r2_bd_df[:20].index)+list(r2_bd_2_df[:20].index)+list(r2_bd_3_df[:20].index)))

len(set(first_3_bd_samps_correlated_feats_list + feats_extracted_from_10000_samp))

20

In [40]:
feats_extracted_from_10000_samp = [int(x[5:]) for x in feats_extracted_from_10000_samp_r2]

In [18]:
ma_first_3 = madelon_analyzer(samp_3000_bd_df[:2000], samp_3000_bd_df[2000:2600])
ma_first_3_benchmark = ma_first_3.train_val_scorer_df_maker(first_3_bd_samps_correlated_feats_list,\
                                                                  'first_3_bd')
ma_first_3_benchmark

first_3_bd
train_score    0.685500
val_score      0.523333

In [19]:
ma_first_3_6506 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
%time ma_first_3_benchmark_6506 = ma_first_3_6506.train_val_scorer_df_maker(first_3_bd_samps_correlated_feats_list,\
                                                                  'first_3_bd_6506')
ma_first_3_benchmark_6506

CPU times: user 1.19 s, sys: 20 ms, total: 1.21 s
Wall time: 2.93 s


first_3_bd_6506
train_score         0.678600
val_score           0.518592

In [20]:
# ma_6506 = madelon_analyzer(samp_6506_bd_df.iloc[:200,500:], samp_6506_bd_df.iloc[500:600,500:])

# %time r2_bd_6506_df = ma_6506.mean_r2_for_all_features(samp_6506_bd_df.iloc[:500,:], DecisionTreeRegressor)

# r2_bd_6506_df = r2_bd_6506_df.sort_values('r2_score', ascending=False)

# pd.to_pickle(r2_bd_6506_df, 'mean_r2_bd_sample_6506.p')
# r2_bd_6506_df = pd.read_pickle('mean_r2_bd_sample_6506.p')

In [28]:
X = samp_6506_bd_df.iloc[:,1:-1]
y = samp_6506_bd_df['target']

skb_20 = SelectKBest(k=20)
skb_20.fit(X, y)
skb_20_feats = np.where(skb_20.get_support())[0]
skb_20_cols = list(X[skb_20_feats].head().columns)

skb_20_cols_new = sorted([int(x[-3:]) for x in skb_20_cols])

In [22]:
pd.DataFrame([sorted(first_3_bd_samps_correlated_feats_list), skb_20_cols_new], \
             index=['first_3_bd_samps','skb_20']).T

first_3_bd_samps  skb_20
0                257     160
1                269     257
2                308     269
3                315     315
4                336     336
5                341     341
6                395     395
7                504     490
8                526     504
9                639     681
10               681     701
11               701     724
12               724     736
13               736     769
14               769     808
15               808     823
16               829     829
17               867     849
18               920     867
19               956     920

In [ ]:
benchmarks_df = pd.merge(ma_first_3_benchmark, ma_first_3_benchmark_6506, left_index=True, right_index=True)

In [24]:
benchmarks_df

first_3_bd  first_3_bd_6506
train_score    0.685500         0.678600
val_score      0.523333         0.518592

In [25]:
ma_skb_6506 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_skb_benchmark_6506 = ma_skb_6506.train_val_scorer_df_maker(skb_20_cols_new,\
                                                                  'skb_20_feats')
ma_skb_benchmark_6506

skb_20_feats
train_score       0.67820
val_score         0.49668

In [26]:
skb_100 = SelectKBest(k=100)
skb_100.fit(X, y)
skb_100_feats = np.where(skb_100.get_support())[0]
skb_100_cols = list(X[skb_100_feats].head().columns)
skb_100_cols_new = sorted([int(x[-3:]) for x in skb_100_cols])
ma_skb_100 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_skb_100_scores = ma_skb_100.train_val_scorer_df_maker(skb_100_cols_new,\
                                                               'skb_100_feats')
ma_skb_100_scores

skb_100_feats
train_score       0.708800
val_score         0.552457

In [29]:
sfm = SelectFromModel(LassoCV(), threshold='mean')
sfm.fit(X, y)
sfm_feats = np.where(sfm.get_support())[0]
sfm_cols = list(X[sfm_feats].head().columns)
sfm_ma = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_sfm_scores = sfm_ma.train_val_scorer_df_maker(sfm_cols, 'sfm_feats')
ma_sfm_scores

sfm_feats
train_score   0.857000
val_score     0.762284

In [30]:
sfm_feats

array([269, 681, 701, 808, 829])

In [27]:
all_benchmark_feats = list(set((first_3_bd_samps_correlated_feats_list + skb_20_cols_new)))

In [28]:
# This shows that all 20 of the columns I selected are included in the top 100 seleced from 
benchmark_plus_skbtop100 = list(set(all_benchmark_feats + skb_100_cols_new))

In [29]:
ma_103 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_103_scores = ma_103.train_val_scorer_df_maker(benchmark_plus_skbtop100, 'benchmark_plus_skbtop100')
ma_103_scores

benchmark_plus_skbtop100
train_score                  0.712200
val_score                    0.538513

In [30]:
# con_103, cur_103 = con_cur_to_class_db()
# cur_103.execute("SELECT {}, target FROM madelon LIMIT 100000".format(feats_103))
# results_103 = cur_103.fetchall()
# con_103.close()
# results_103_100000rows_df = pd.DataFrame(results_103)
# pd.to_pickle(results_103_100000rows_df, 'Data/results_103_100000rows_df.p')

In [31]:
results_103_100000rows_df.head()

feat_004  feat_010  feat_061  feat_066  feat_078  feat_094  feat_100  \
0 -0.883567  1.684800  0.031787 -0.613555  0.365499 -0.090137  0.645975   
1  0.721962 -0.736924 -0.524250 -1.236556  1.372517 -0.488248 -0.891683   
2  1.090644  0.227951 -0.357068 -1.204669  1.244378 -0.519236  1.668885   
3 -0.137440  0.282772  1.395812  0.548668  0.177660  1.876143  0.157389   
4 -0.817359  0.058184  0.595671  0.323599  1.235149  0.774004  0.639562   

   feat_104  feat_107  feat_124   ...    feat_941  feat_948  feat_954  \
0  0.066871  1.691925  1.868239   ...   -0.282699  0.248959 -1.571380   
1  0.197044 -0.473649 -0.967551   ...    0.598059  1.636510 -1.492728   
2 -0.830607 -2.515024 -0.450493   ...    1.416518 -0.254385  0.833637   
3  0.394322 -0.460903 -0.995456   ...    1.412139 -0.446674 -1.388820   
4  0.557761 -1.513044  0.280326   ...    1.089422 -1.235951 -1.386048   

   feat_956  feat_969  feat_970  feat_977  feat_995  feat_996  target  
0  1.072009  1.171244  0.542077  0.375039 -1.363639 -0.737410       1  
1 -3.566485 -0.763321  0.415703  1.027809  1.072134 -0.110998       1  
2 -2.919125 -1.831422 -0.237104  0.877890 -0.302300  1.435411       1  
3  0.065801 -1.939555  0.870874  1.629417 -0.057647 -1.176244       1  
4  1.349128 -0.379900 -0.875226  0.609906  0.617287  0.691262       1  

[5 rows x 104 columns]

In [33]:
X_103 = results_103_100000rows_df.T[:-1].T
y_103 = results_103_100000rows_df.T[-1:].T

sfm_103 = SelectFromModel(LassoCV(), threshold='mean')
sfm_103.fit(X_103, y_103)
sfm_103_feats = np.where(sfm_103.get_support())[0]
sfm_103_cols = list(X_103.head(1).columns)
sfm_103_ma = madelon_analyzer(results_103_100000rows_df[:10000], samp_3000_bd_df)
ma_sfm_103_scores = sfm_103_ma.train_val_scorer_df_maker(sfm_103_cols, 'sfm103_feats_100000rows')
ma_sfm_103_scores

/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:1094: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


sfm103_feats_100000rows
train_score                 0.766900
val_score                   0.626667

In [34]:
sfm2 = SelectFromModel(LassoCV(), threshold='mean')
sfm2.fit(X[all_benchmark_feats], y)
sfm2_feats = np.where(sfm2.get_support())[0]
sfm2_cols = list(X[sfm2_feats].head().columns)
sfm2_ma = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_sfm2_scores = sfm2_ma.train_val_scorer_df_maker(sfm2_cols, 'sfm2_feats')
ma_sfm2_scores

sfm2_feats
train_score    0.699000
val_score      0.508632

In [35]:
val_df = samp_3000_bd_df.T[1:].T
train_df = samp_6506_bd_df.T[1:].T

In [5]:
benchmark_plus_skbtop100_features = [x for x in pd.read_pickle('Data/results_103_100000rows_df.p').columns[:-1]]

In [36]:
%run madelon_analyzer

RFE = RFE(DecisionTreeClassifier(), n_features_to_select=20)

In [48]:
RFE.fit(train_df[benchmark_plus_skbtop100_features], train_df.T[-1:].T)

/opt/conda/lib/python3.6/site-packages/sklearn/utils/validation.py:547: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


RFE(estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'),
  n_features_to_select=20, step=1, verbose=0)

In [49]:
RFE_feats = np.where(RFE.get_support())[0]
RFE_cols = list(train_df[benchmark_plus_skbtop100_features][RFE_feats].columns)

In [53]:
ma_rfe = madelon_analyzer(train_df, val_df)
RFE_cols_score_df = ma_rfe.train_val_scorer_df_maker(RFE_cols, 'RFE_cols', \
                                                     params={'n_neighbors':[1,3,5,7,9,11,13,15,17,19,21]})
RFE_cols_score_df.T

train_score  val_score
RFE_cols     0.785736      0.736

In [52]:
ma_rfe.best_params_

{'n_neighbors': 15}

In [57]:
# # In this cell, I merged all the benchmarks into one dataframe

# benchmarks_df = pd.merge(ma_first_3_benchmark, ma_first_3_benchmark_6506, left_index=True, right_index=True)
# benchmarks_df = pd.merge(benchmarks_df, ma_skb_benchmark_6506, left_index=True, right_index=True)
# benchmarks_df = pd.merge(benchmarks_df, ma_sfm_scores, left_index=True, right_index=True)
# benchmarks_df = pd.merge(benchmarks_df, ma_sfm2_scores, left_index=True, right_index=True)
# benchmarks_df = pd.merge(benchmarks_df, ma_skb_100_scores, left_index=True, right_index=True)
# benchmarks_df = pd.merge(benchmarks_df, ma_103_scores, left_index=True, right_index=True)
# multi_benchmarks_df = benchmarks_df.T.sort_values('val_score', ascending=False)
# multi_benchmarks_df = pd.merge(multi_benchmarks_df.T, RFE_cols_score_df, left_index=True, right_index=True)
# multi_benchmarks_df = multi_benchmarks_df.T.sort_values('val_score', ascending=False)
# multi_benchmarks_df['description'] = ['Top 5 using SelectFromModel with LassoCV, threshold=mean',
#                                       'Top 20 using RFE',
#                                      'Top 100 using SelectKBest', 'SKB top 100 plus top 20',
#                                       'Top 20 from the small samples', 
#                                      'Using the same list from Top 20 from the small samples but on representitive sized sample',
#                                      'SFM, on only the top 20 from small samples features', 'SKB top 20']
# pd.to_pickle(multi_benchmarks_df, 'Data/multi_benchmarks_df.p')

In [59]:
pd.to_pickle(multi_benchmarks_df, 'Data/multi_benchmarks_df.p')
multi_benchmarks_df = pd.read_pickle('Data/multi_benchmarks_df.p')

In [34]:
multi_benchmarks_df['description'] = ['5 features identified by SFM', '20 features from RFE', \
                                      'SelectKBest top 100', '20 columns identified using regression, plus the skb\
                                      top 100', '20 columns identified using regression tested on 3000 sample',
                                     '20 columns identified using regression tested on different 6506 sample',
                                     'Used SFM on only the top 20 features identified using regression', 'SKB 20\
                                     features']

In [35]:
multi_benchmarks_df

train_score val_score  \
sfm_feats                      0.857  0.762284   
RFE_cols                    0.785736     0.736   
skb_100_feats                 0.7088  0.552457   
benchmark_plus_skbtop100      0.7122  0.538513   
first_3_bd                    0.6855  0.523333   
first_3_bd_6506               0.6786  0.518592   
sfm2_feats                     0.699  0.508632   
skb_20_feats                  0.6782   0.49668   

                                                                description  
sfm_feats                                      5 features identified by SFM  
RFE_cols                                               20 features from RFE  
skb_100_feats                                           SelectKBest top 100  
benchmark_plus_skbtop100  20 columns identified using regression, plus t...  
first_3_bd                20 columns identified using regression tested ...  
first_3_bd_6506           20 columns identified using regression tested ...  
sfm2_feats                Used SFM on only the top 20 features identifie...  
skb_20_feats              SKB 20                                     fea...

## Thoughts after benchmarking and initial feature selection

Using a naive KNeighborsClassifier, I was able to score a 0.762284 for my validation set. From three random 1% samples of the data, I was able to identify 20 colinear features, which scored 0.518592 when I validated it on a random sample of 6506 rows, which is a representative sample of the population with a 90% confidence level and a 1% margin of error. Since the two clases are balanced, this is not an effective model.

Using SelectKBest to get 100 features, I was able to score higher on my validation set than when I looked for 20 features. This tells me that I was probably missing some of the important features in the set of 20 using this method. Although, I descovered that my benchmark list of 20 included 3 features that were not in the SKB top 100 list. So I added them, and will be using all 103 features.

---

In [6]:
benchmark_plus_skbtop100_features = [x for x in pd.read_pickle('Data/results_103_100000rows_df.p').columns[:-1]]
benchmark_plus_skbtop100_features.append('target')

In [7]:
X_r2_103_feats = results_103_100000rows_df[benchmark_plus_skbtop100_features]


In [ ]:
ma_103_100000 = madelon_analyzer(results_103_100000rows_df[:60000], samp_3000_bd_df)

bd_103_100000 = ma_103_100000.mean_r2_for_all_features(X_r2_103_feats, DecisionTreeRegressor)
bd_103_100000_df = bd_103_100000.sort_values('r2_score', ascending=False)
bd_103_100000_df

In [ ]:
pd.to_pickle(bd_103_100000_df, 'mean_r2_bd_103_100000_df.p')
bd_103_100000_df = pd.read_pickle('Data/mean_r2_bd_103_100000_df.p')
bd_103_100000_df[:25]

In [4]:
multi_benchmarks_df

train_score val_score
sfm_feats                      0.857  0.762284
RFE_cols                    0.785736     0.736
skb_100_feats                 0.7088  0.552457
benchmark_plus_skbtop100      0.7122  0.538513
first_3_bd                    0.6855  0.523333
first_3_bd_6506               0.6786  0.518592
sfm2_feats                     0.699  0.508632
skb_20_feats                  0.6782   0.49668

## Feature Importance

Using the madelon_analyzer, I took the top 103 features list and GridSearched a few models including RandomForestClassifier and GradientBoostingClassifier which did improve my model score from benchmark. 

Using the madelon_analyzer I created, I've been able to score better than benchmark on a couple of occasion. I found that I was able to improve scores by using the ```madelon_analyzer.list_top_dipped_feats(model=KNeighborsClassifier, params='n_neighbors':range(1,20,2), random=True, noise=True)``` method and then repeating the process with a list of features that has the 'noisey' features removed. Below are the graphs that get output, showing the features that were removed marked in blue. By repeating this process several times, I was able to improve my validation score to 0.8185.

In [34]:
rf_params = {
    'n_estimators':[10,50,100,200],
    'max_features':['auto','log2'],
}

ma_fi = madelon_analyzer(train_df, val_df)
%time ma_fi_scores_103 = ma_fi.train_val_scorer_df_maker(benchmark_plus_skbtop100_features, '103 features, RFC',\
                                                     RandomForestClassifier, rf_params)
ma_fi_scores_103

CPU times: user 10 s, sys: 150 ms, total: 10.2 s
Wall time: 40.3 s


103 features, RFC
train_score              1.000
val_score                0.755

In [35]:
gbparams = {
    'learning_rate': [0.1,0.5,1], 
    'n_estimators': [100,200], 
    'max_depth': [2,4], 
}

ma_gb = madelon_analyzer(train_df, val_df)
%time ma_gb_scores_103 = ma_fi.train_val_scorer_df_maker(benchmark_plus_skbtop100_features, '103 features, GBC',\
                                                     GradientBoostingClassifier, gbparams)
ma_gb_scores_103

CPU times: user 12.6 s, sys: 80 ms, total: 12.7 s
Wall time: 2min 5s


103 features, GBC
train_score           0.932524
val_score             0.724667

In [37]:
val_df['target'].value_counts()

0.0    1511
1.0    1489
Name: target, dtype: int64

In [77]:
ma_fi = madelon_analyzer(results_103_100000rows_df[:10000], results_103_100000rows_df[10000:12000])

In [ ]:
fi_scores_bd_RFE_cols = ma_fi.list_top_dipped_feats(bd_third_best_feats,\
                                            KNeighborsClassifier,\
                                            params={},\
                                            noise=True,\
                                            random=True,\
                                            n_feats_return=2)

In [ ]:
good_cols = ma_fi.removerizer(bd_best_feats, fi_scores_bd_RFE_cols)
good_cols

In [ ]:
good_cols_df = ma_fi.train_val_scorer_df_maker(good_cols, 'with_noise_removal', params={'n_neighbors':range(3,20,2)})
good_cols_df

In [104]:
gc_df = ma_fi.train_val_scorer_df_maker(bd_best_feats, 'with_noise_removal', params={'n_neighbors':range(3,20,2)})
gc_df

with_noise_removal
train_score              0.8582
val_score                0.8185

# Results and Top 5 best score:

After reducing the list of features using the madelon_analyzer, I tried every combination of 5 features and found that the best scoring set was ```['feat_681', 'feat_829', 'feat_308', 'feat_724', 'feat_956']```.

These Top 5 features when tested with KNN were only marginally better with a validation score of 0.8205. 

In [143]:
bd_ma_top_scores = madelon_analyzer(results_103_100000rows_df[:10000], results_103_100000rows_df[10000:12000])

dfs = []

for i, l in enumerate(bd_top_scores_list_of_lists):
    if i == 0:
        df = bd_ma_top_scores.train_val_scorer_df_maker(l, 'Top 5 Score', \
                                                 params={'n_neighbors':range(1,20,2)})
    else:
        df = bd_ma_top_scores.train_val_scorer_df_maker(l, 'Number {} score'.format(i), \
                                                 params={'n_neighbors':range(1,20,2)})
    dfs.append(df)

In [144]:
df_merge = pd.merge(dfs[0], dfs[1], left_index=True, right_index=True)
df_merge = pd.merge(df_merge, dfs[2], left_index=True, right_index=True)
top_scores_df = pd.merge(df_merge, dfs[3], left_index=True, right_index=True)

top_scores_df

Top 5 Score  Number 1 score  Number 2 score  Number 3 score
train_score       0.8669          0.8582          0.8414          0.8205
val_score         0.8205          0.8185          0.7985          0.7845

In [130]:
bd_brute_force_combos_df = \
bd_ma_top_scores.brute_force_feature_combination_score_generator(bd_second_best_feats, \
                                                                 model=KNeighborsClassifier,\
                                                                 n_feats=5, \
                                                                 params={'n_neighbors':[3,20,2]})

Number of combinations: 252


In [131]:
bd_brute_force_combos_df.iloc[:,1:].head()

train_score val_score
['feat_681', 'feat_829', 'feat_308', 'feat_724'...      0.8897     0.821
['feat_681', 'feat_920', 'feat_724', 'feat_736'...      0.8949     0.815
['feat_829', 'feat_701', 'feat_308', 'feat_769'...      0.8945    0.8145
['feat_681', 'feat_920', 'feat_808', 'feat_736'...      0.8289    0.8145
['feat_920', 'feat_308', 'feat_769', 'feat_808'...      0.8934     0.814

In [5]:
bd_top_5_brute_f

['feat_308', 'feat_681', 'feat_724', 'feat_829', 'feat_956']

In [156]:
rf_params = {
    'n_estimators':[10,50,100,200],
    'max_features':['auto','log2'],
}

bd_ma_five_scores = bd_ma_top_scores.train_val_scorer_df_maker(bd_top_5_brute_f, 'top 5 features, RFC',\
                                                     RandomForestClassifier, rf_params)
bd_ma_five_scores

top 5 features, RFC
train_score               1.0000
val_score                 0.8055

In [7]:
X_train = results_103_100000rows_df[bd_top_5_brute_f][:10000]
X_val = results_103_100000rows_df[bd_top_5_brute_f][10000:13000]
y_train = results_103_100000rows_df['target'][:10000]
y_val = results_103_100000rows_df['target'][10000:13000]

knc_params = {
    'n_neighbors':range(3,30,2),
    'p':[1,2]
    
}

rf_params = {
    'criterion':['gini','entropy'],
    'max_depth':[None,2,3,4,5,20,100],
    
}

knc_roc = GridSearchCV(KNeighborsClassifier(n_jobs=-1), knc_params)
rf_roc = GridSearchCV(RandomForestClassifier(n_jobs=-1), rf_params)

rf_roc.fit(X_train, y_train)
knc_roc.fit(X_train, y_train)

y_preds = knc_roc.predict(X_val)
y_proba = knc_roc.predict_proba(X_val)

y_preds_rf = rf_roc.predict(X_val)
y_proba_rf = rf_roc.predict_proba(X_val)

In [8]:
roc_auc_score(y_val, y_preds)

0.81061276445575925

In [9]:
roc_auc_score(y_val, y_preds_rf)

0.77515532560486866

In [10]:
rf_roc.best_params_, knc_roc.best_params_

({'criterion': 'entropy', 'max_depth': None}, {'n_neighbors': 7, 'p': 2})

In [11]:
X_train = results_103_100000rows_df[bd_best_feats][:10000]
X_val = results_103_100000rows_df[bd_best_feats][10000:13000]
y_train = results_103_100000rows_df['target'][:10000]
y_val = results_103_100000rows_df['target'][10000:13000]

knc_params = {
    'n_neighbors':range(3,30,2),
    'p':[1,2]
    
}

rf_params = {
    'criterion':['gini','entropy'],
    'max_depth':[None,2,3,4,5,20,100],
    
}

knc_roc = GridSearchCV(KNeighborsClassifier(n_jobs=-1), knc_params)
rf_roc = GridSearchCV(RandomForestClassifier(n_jobs=-1), rf_params)

rf_roc.fit(X_train, y_train)
knc_roc.fit(X_train, y_train)

y_preds = knc_roc.predict(X_val)
y_proba = knc_roc.predict_proba(X_val)

y_preds_rf = rf_roc.predict(X_val)
y_proba_rf = rf_roc.predict_proba(X_val)

In [12]:
roc_auc_score(y_val, y_preds)

0.81193679329016499

In [13]:
roc_auc_score(y_val, y_preds_rf)

0.79497531279570088

In [14]:
rf_roc.best_params_, knc_roc.best_params_

({'criterion': 'entropy', 'max_depth': 20}, {'n_neighbors': 7, 'p': 2})

In [21]:
X_train = results_103_100000rows_df.iloc[:10000,:-1]
X_val = results_103_100000rows_df.iloc[10000:13000,:-1]
y_train = results_103_100000rows_df['target'][:10000]
y_val = results_103_100000rows_df['target'][10000:13000]

knc_params = {
    'n_neighbors':range(3,30,2),
    'p':[1,2]
    
}

rf_params = {
    'criterion':['gini','entropy'],
    'max_depth':[None,2,3,4,5,20,100],
    
}

knc_roc = GridSearchCV(KNeighborsClassifier(n_jobs=-1), knc_params)
rf_roc = GridSearchCV(RandomForestClassifier(n_jobs=-1), rf_params)

rf_roc.fit(X_train, y_train)
knc_roc.fit(X_train, y_train)

y_preds = knc_roc.predict(X_val)
y_proba = knc_roc.predict_proba(X_val)

y_preds_rf = rf_roc.predict(X_val)
y_proba_rf = rf_roc.predict_proba(X_val)

In [22]:
roc_auc_score(y_val, y_preds)

0.75865518849077151

In [23]:
roc_auc_score(y_val, y_preds_rf)

0.72461200266139136

In [24]:
rf_roc.best_params_, knc_roc.best_params_

({'criterion': 'entropy', 'max_depth': None}, {'n_neighbors': 19, 'p': 1})

In [31]:
X_train = results_103_100000rows_df[sfm_cols][:10000]
X_val = results_103_100000rows_df[sfm_cols][10000:13000]
y_train = results_103_100000rows_df['target'][:10000]
y_val = results_103_100000rows_df['target'][10000:13000]

knc_params = {
    'n_neighbors':range(3,30,2),
    'p':[1,2]
    
}

rf_params = {
    'criterion':['gini','entropy'],
    'max_depth':[None,2,3,4,5,20,100],
    
}

knc_roc = GridSearchCV(KNeighborsClassifier(n_jobs=-1), knc_params)
rf_roc = GridSearchCV(RandomForestClassifier(n_jobs=-1), rf_params)

rf_roc.fit(X_train, y_train)
knc_roc.fit(X_train, y_train)

y_preds = knc_roc.predict(X_val)
y_proba = knc_roc.predict_proba(X_val)

y_preds_rf = rf_roc.predict(X_val)
y_proba_rf = rf_roc.predict_proba(X_val)

In [32]:
roc_auc_score(y_val, y_preds), roc_auc_score(y_val, y_preds_rf)

(0.78926963298311836, 0.76202592856466644)

In [33]:
rf_roc.best_params_, knc_roc.best_params_

({'criterion': 'gini', 'max_depth': 20}, {'n_neighbors': 5, 'p': 2})

--- 
 
---